File: interview-coach-backend/schemas/interview.py
from datetime import datetime
from typing import Literal
from pydantic import BaseModel, Field


Difficulty = Literal["easy", "medium", "hard"]
QuestionType = Literal["behavioral", "technical", "project_based", "system_design"]
InterviewMode = Literal["general", "resume_based", "job_description_based"]



class StartInterviewRequest(BaseModel):
    role: str = Field(..., min_length=2, max_length=255, examples=["Backend Engineer Intern"])
    difficulty: Difficulty = "medium"
    mode: InterviewMode = "general"
    resume_text: str | None = Field(None, max_length=20000)
    job_description_text: str | None = Field(None, max_length=10000)


class SubmitAnswerRequest(BaseModel):
    session_id: str
    question_id: str
    answer: str = Field(..., min_length=1, max_length=8000)


class EndInterviewRequest(BaseModel):
    session_id: str


class QuestionOut(BaseModel):
    id: str
    order_index: int
    question: str
    question_type: QuestionType

    class Config:
        from_attributes = True


class StartInterviewResponse(BaseModel):
    session_id: str
    role: str
    difficulty: Difficulty
    mode: InterviewMode
    question: QuestionOut


class AnswerEvaluation(BaseModel):
    score: float = Field(..., ge=0, le=10)
    feedback: str
    ideal_answer_hints: list[str] = Field(default_factory=list)


class SubmitAnswerResponse(BaseModel):
    evaluation: AnswerEvaluation
    next_question: QuestionOut | None = None
    is_complete: bool = False
    questions_asked: int
    max_questions: int


class QuestionDetail(BaseModel):
    id: str
    order_index: int
    question: str
    question_type: QuestionType
    answer: str | None
    score: float | None
    feedback: str | None

    class Config:
        from_attributes = True


class EndInterviewResponse(BaseModel):
    session_id: str
    overall_score: float
    technical_score: float
    communication_score: float
    summary: str
    recommendations: list[str] = Field(default_factory=list)
    questions: list[QuestionDetail]


class InterviewSessionSummary(BaseModel):
    id: str
    role: str
    difficulty: Difficulty
    mode: InterviewMode
    status: str
    overall_score: float | None
    technical_score: float | None
    communication_score: float | None
    created_at: datetime
    completed_at: datetime | None
    question_count: int

    class Config:
        from_attributes = True


class InterviewHistoryResponse(BaseModel):
    total: int
    items: list[InterviewSessionSummary]


class InterviewSessionDetail(BaseModel):
    id: str
    role: str
    difficulty: Difficulty
    mode: InterviewMode
    status: str
    overall_score: float | None
    technical_score: float | None
    communication_score: float | None
    summary: str | None
    recommendations: list[str] = Field(default_factory=list)
    created_at: datetime
    completed_at: datetime | None
    questions: list[QuestionDetail]

    class Config:
        from_attributes = True

File: interview-coach-frontend/src/types/interview.ts
export type Difficulty = "easy" | "medium" | "hard";
export type QuestionType = "behavioral" | "technical" | "project_based" | "system_design";
export type InterviewMode = "general" | "resume_based" | "job_description_based";

export interface StartInterviewRequest {
  role: string;
  difficulty: Difficulty;
  mode: InterviewMode;
  resume_text?: string | null;
  job_description_text?: string | null;
}

export interface QuestionOut {
  id: string;
  order_index: number;
  question: string;
  question_type: QuestionType;
}

export interface StartInterviewResponse {
  session_id: string;
  role: string;
  difficulty: Difficulty;
  mode: InterviewMode;
  question: QuestionOut;
}

export interface SubmitAnswerRequest {
  session_id: string;
  question_id: string;
  answer: string;
}

export interface AnswerEvaluation {
  score: number;
  feedback: string;
  ideal_answer_hints: string[];
}

export interface SubmitAnswerResponse {
  evaluation: AnswerEvaluation;
  next_question: QuestionOut | null;
  is_complete: boolean;
  questions_asked: number;
  max_questions: number;
}

export interface QuestionDetail {
  id: string;
  order_index: number;
  question: string;
  question_type: QuestionType;
  answer: string | null;
  score: number | null;
  feedback: string | null;
}

export interface EndInterviewResponse {
  session_id: string;
  overall_score: number;
  technical_score: number;
  communication_score: number;
  summary: string;
  recommendations: string[];
  questions: QuestionDetail[];
}

export interface InterviewSessionSummary {
  id: string;
  role: string;
  difficulty: Difficulty;
  mode: InterviewMode;
  status: "active" | "completed";
  overall_score: number | null;
  technical_score: number | null;
  communication_score: number | null;
  created_at: string;
  completed_at: string | null;
  question_count: number;
}

export interface InterviewHistoryResponse {
  total: number;
  items: InterviewSessionSummary[];
}

export interface InterviewSessionDetail {
  id: string;
  role: string;
  difficulty: Difficulty;
  mode: InterviewMode;
  status: "active" | "completed";
  overall_score: number | null;
  technical_score: number | null;
  communication_score: number | null;
  summary: string | null;
  recommendations: string[];
  created_at: string;
  completed_at: string | null;
  questions: QuestionDetail[];
}